File: src/knowledge_mcp/storage/atoms.py
"""Atom file storage management."""

from __future__ import annotations

import json
from pathlib import Path
from typing import TYPE_CHECKING

import yaml

from ..config import Config, get_config
from ..models.atom import Atom

if TYPE_CHECKING:
    pass


def _multiline_str_representer(dumper: yaml.Dumper, data: str) -> yaml.ScalarNode:
    """Represent multiline strings using literal block style (|)."""
    if "\n" in data:
        # Use literal block style for multiline strings
        return dumper.represent_scalar("tag:yaml.org,2002:str", data, style="|")
    return dumper.represent_scalar("tag:yaml.org,2002:str", data)


class _MultilineDumper(yaml.SafeDumper):
    """Custom YAML dumper that uses literal block style for multiline strings."""

    pass


_MultilineDumper.add_representer(str, _multiline_str_representer)


class AtomStorage:
    """Manages atom file storage."""

    def __init__(self, config: Config | None = None) -> None:
        self.config = config or get_config()

    def _yaml_path(self, atom_id: str) -> Path:
        """Get the YAML path for an atom file."""
        return self.config.atoms_path / f"{atom_id}.yaml"

    def _json_path(self, atom_id: str) -> Path:
        """Get the legacy JSON path for an atom file."""
        return self.config.atoms_path / f"{atom_id}.json"

    def save(self, atom: Atom) -> Path:
        """Save atom to disk in YAML format.

        If a legacy JSON file exists, it is deleted after successful YAML write.
        """
        self.config.ensure_dirs()

        yaml_path = self._yaml_path(atom.id)
        json_path = self._json_path(atom.id)

        # Convert to dict and serialize as YAML
        data = atom.model_dump()
        with open(yaml_path, "w", encoding="utf-8") as f:
            yaml.dump(
                data,
                f,
                Dumper=_MultilineDumper,
                default_flow_style=False,
                allow_unicode=True,
                sort_keys=False,
            )

        # Clean up legacy JSON file if it exists
        if json_path.exists():
            json_path.unlink(missing_ok=True)

        return yaml_path

    def load(self, atom_id: str) -> Atom | None:
        """Load atom from disk.

        Tries YAML first, falls back to JSON for backward compatibility.
        Returns None if atom doesn't exist.
        """
        yaml_path = self._yaml_path(atom_id)
        json_path = self._json_path(atom_id)

        # Try YAML first
        if yaml_path.exists():
            with open(yaml_path, encoding="utf-8") as f:
                data = yaml.safe_load(f)
            return Atom.model_validate(data)

        # Fall back to JSON
        if json_path.exists():
            with open(json_path, encoding="utf-8") as f:
                data = json.load(f)
            return Atom.model_validate(data)

        return None

    def delete(self, atom_id: str) -> bool:
        """Delete atom files from disk (both YAML and JSON).

        Returns True if any file was deleted, False otherwise.
        """
        yaml_path = self._yaml_path(atom_id)
        json_path = self._json_path(atom_id)

        deleted = False

        if yaml_path.exists():
            yaml_path.unlink()
            deleted = True
        if json_path.exists():
            json_path.unlink()
            deleted = True

        return deleted

    def exists(self, atom_id: str) -> bool:
        """Check if an atom file exists (YAML or JSON)."""
        return self._yaml_path(atom_id).exists() or self._json_path(atom_id).exists()

    def list_all_ids(self) -> list[str]:
        """List all atom IDs in storage."""
        if not self.config.atoms_path.exists():
            return []

        # Use a set to deduplicate IDs (in case both .yaml and .json exist)
        id_set: set[str] = set()

        for path in self.config.atoms_path.iterdir():
            if path.is_dir():
                continue
            name = path.name
            if not name.startswith("K-"):
                continue
            if name.endswith(".yaml"):
                id_set.add(name.removesuffix(".yaml"))
            elif name.endswith(".json"):
                id_set.add(name.removesuffix(".json"))

        return sorted(id_set)

File: storage/atoms.go
package storage

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"gopkg.in/yaml.v3"

	"github.com/sephriot/knowledge-mcp/config"
	"github.com/sephriot/knowledge-mcp/models"
)

// AtomStorage manages atom file storage.
type AtomStorage struct {
	config *config.Config
}

// NewAtomStorage creates a new atom storage.
func NewAtomStorage(cfg *config.Config) *AtomStorage {
	if cfg == nil {
		cfg = config.GetConfig()
	}
	return &AtomStorage{config: cfg}
}

// getAtomPathYAML returns the YAML path for an atom file.
func (s *AtomStorage) getAtomPathYAML(atomID string) string {
	return filepath.Join(s.config.AtomsPath(), fmt.Sprintf("%s.yaml", atomID))
}

// getAtomPathJSON returns the legacy JSON path for an atom file.
func (s *AtomStorage) getAtomPathJSON(atomID string) string {
	return filepath.Join(s.config.AtomsPath(), fmt.Sprintf("%s.json", atomID))
}

// Save saves an atom to disk in YAML format.
// If a legacy JSON file exists, it is deleted after successful YAML write.
func (s *AtomStorage) Save(atom *models.Atom) (string, error) {
	if err := s.config.EnsureDirs(); err != nil {
		return "", fmt.Errorf("failed to create directories: %w", err)
	}

	yamlPath := s.getAtomPathYAML(atom.ID)
	jsonPath := s.getAtomPathJSON(atom.ID)

	data, err := yaml.Marshal(atom)
	if err != nil {
		return "", fmt.Errorf("failed to marshal atom to YAML: %w", err)
	}

	if err := os.WriteFile(yamlPath, data, 0644); err != nil {
		return "", fmt.Errorf("failed to write atom file: %w", err)
	}

	// Clean up legacy JSON file if it exists
	if _, err := os.Stat(jsonPath); err == nil {
		os.Remove(jsonPath) // Best effort, ignore errors
	}

	return yamlPath, nil
}

// Load loads an atom from disk.
// Tries YAML first, falls back to JSON for backward compatibility.
func (s *AtomStorage) Load(atomID string) (*models.Atom, error) {
	yamlPath := s.getAtomPathYAML(atomID)
	jsonPath := s.getAtomPathJSON(atomID)

	// Try YAML first
	if data, err := os.ReadFile(yamlPath); err == nil {
		var atom models.Atom
		if err := yaml.Unmarshal(data, &atom); err != nil {
			return nil, fmt.Errorf("failed to unmarshal YAML atom: %w", err)
		}
		return &atom, nil
	}

	// Fall back to JSON
	data, err := os.ReadFile(jsonPath)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, fmt.Errorf("failed to read atom file: %w", err)
	}

	var atom models.Atom
	if err := json.Unmarshal(data, &atom); err != nil {
		return nil, fmt.Errorf("failed to unmarshal JSON atom: %w", err)
	}

	return &atom, nil
}

// Delete deletes an atom file from disk (both YAML and JSON versions).
func (s *AtomStorage) Delete(atomID string) (bool, error) {
	yamlPath := s.getAtomPathYAML(atomID)
	jsonPath := s.getAtomPathJSON(atomID)

	yamlExists := false
	jsonExists := false

	if _, err := os.Stat(yamlPath); err == nil {
		yamlExists = true
	}
	if _, err := os.Stat(jsonPath); err == nil {
		jsonExists = true
	}

	if !yamlExists && !jsonExists {
		return false, nil
	}

	if yamlExists {
		if err := os.Remove(yamlPath); err != nil {
			return false, fmt.Errorf("failed to delete YAML atom file: %w", err)
		}
	}
	if jsonExists {
		if err := os.Remove(jsonPath); err != nil {
			return false, fmt.Errorf("failed to delete JSON atom file: %w", err)
		}
	}

	return true, nil
}

// Exists checks if an atom file exists (YAML or JSON).
func (s *AtomStorage) Exists(atomID string) bool {
	yamlPath := s.getAtomPathYAML(atomID)
	jsonPath := s.getAtomPathJSON(atomID)

	if _, err := os.Stat(yamlPath); err == nil {
		return true
	}
	if _, err := os.Stat(jsonPath); err == nil {
		return true
	}
	return false
}

// ListAllIDs lists all atom IDs in storage.
func (s *AtomStorage) ListAllIDs() ([]string, error) {
	atomsPath := s.config.AtomsPath()

	if _, err := os.Stat(atomsPath); os.IsNotExist(err) {
		return []string{}, nil
	}

	entries, err := os.ReadDir(atomsPath)
	if err != nil {
		return nil, fmt.Errorf("failed to read atoms directory: %w", err)
	}

	// Use a map to deduplicate IDs (in case both .yaml and .json exist)
	idSet := make(map[string]bool)
	for _, entry := range entries {
		if entry.IsDir() {
			continue
		}
		name := entry.Name()
		if !strings.HasPrefix(name, "K-") {
			continue
		}
		if strings.HasSuffix(name, ".yaml") {
			idSet[strings.TrimSuffix(name, ".yaml")] = true
		} else if strings.HasSuffix(name, ".json") {
			idSet[strings.TrimSuffix(name, ".json")] = true
		}
	}

	ids := make([]string, 0, len(idSet))
	for id := range idSet {
		ids = append(ids, id)
	}

	return ids, nil
}
